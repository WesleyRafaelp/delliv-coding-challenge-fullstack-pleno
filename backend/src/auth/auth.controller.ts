import {
  Body,
  Controller,
  Post,
  Req,
  Res,
  UnauthorizedException,
  UseGuards,
  Get,
  HttpException,
  HttpStatus,
} from '@nestjs/common';
import { ApiBearerAuth, ApiTags } from '@nestjs/swagger';
import { LoginDto } from 'src/users/dto/login.dto';
import { AuthService } from './auth.service';
import { RefreshTokenCookie } from './decorators/get-refresh-token.decorator';
import { SetRefreshTokenGuard } from './set-refresh-token.guard';
import { JwtAuthGuard } from './jwt-auth.guard';
import { AuthToken } from './decorators/get-token.decorator';
import { RefreshTokenService } from 'src/refresh-token/refresh-token.service';
import { Response } from 'express';

@ApiTags('auth')
@Controller('auth')
export class AuthController {
  constructor(
    private readonly authService: AuthService,
    private refreshTokenService: RefreshTokenService,
  ) {}

  @Post('login')
  async login(
    @Body() login: LoginDto,
    @Res({ passthrough: true }) res: Response,
    // @RefreshTokenCookie() refresh,
    // @Req() req: Request,
  ) {
    const { accessToken, refreshToken } = await this.authService.login(login);

    res.cookie('REFRESH_TOKEN', refreshToken, {
      httpOnly: true,
      path: '/',
      maxAge: 30 * 24 * 60 * 60 * 1000, // 30 dias
    });
    return { access_token: accessToken };
  }

  @Post('refresh-token')
  @UseGuards(SetRefreshTokenGuard)
  async refreshAccessToken(
    @Req() req,
    @Res({ passthrough: true }) res: Response,
    @RefreshTokenCookie() refresh,
  ) {
    const user = req.user;
    const refreshTokenOld = req.body.refresh_token;
    const validate = await this.refreshTokenService.validateToken(
      refreshTokenOld || refresh.REFRESH_TOKEN,
    );
    if (validate !== true) {
      throw new HttpException(`Token invalid!`, HttpStatus.UNAUTHORIZED);
    }

    const savedToken = await this.refreshTokenService.create(
      refreshTokenOld || refresh.REFRESH_TOKEN,
      user.id,
    );
    if (savedToken !== true) {
      throw new HttpException(`Token invalid!`, HttpStatus.BAD_REQUEST);
    }
    const refreshToken = await this.authService.generateRefreshToken(user.id);
    const { accessToken } = await this.authService.refreshAccessToken(user);
    res.cookie('REFRESH_TOKEN', refreshToken, {
      httpOnly: true,
      path: '/',
      maxAge: 30 * 24 * 60 * 60 * 1000, // 30 dias
    });
    res.cookie('ACCESS_TOKEN', accessToken, {
      path: '/',
      maxAge: 30 * 24 * 60 * 60 * 1000, // 30 dias
    });
    return { access_token: accessToken, refresh_token: refreshToken };
  }

  @Get('session')
  @ApiBearerAuth('JWT-auth')
  @UseGuards(JwtAuthGuard)
  async verifyToken(@AuthToken() token: string): Promise<any> {
    if (!token) throw new UnauthorizedException('Invalid token');
    const isValid = await this.authService.verifyToken(token);

    if (!isValid) {
      throw new UnauthorizedException('Invalid token');
    }

    return {
      message: 'Token is valid',
      data: isValid,
    };
  }

  @Post('logout')
  @ApiBearerAuth('JWT-auth')
  @UseGuards(JwtAuthGuard)
  async logout(@Res({ passthrough: true }) res) {
    res.cookie('REFRESH_TOKEN', null, { maxAge: -1 });
    return { message: 'Logged out successfully' };
  }
}
