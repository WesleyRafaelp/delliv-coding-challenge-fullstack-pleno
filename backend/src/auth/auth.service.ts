import { Injectable, UnauthorizedException } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { UsersService } from 'src/users/users.service';
import { compareSync } from 'bcrypt';
import { JwtPayload } from './interfaces/jwt-payload.interface';
import { LoginDto } from 'src/users/dto/login.dto';
import { BadRequestException } from '@nestjs/common/exceptions';
import { User } from '@prisma/client';

@Injectable()
export class AuthService {
  constructor(
    private readonly userService: UsersService,
    private readonly jwtService: JwtService,
  ) {}

  async validateUser(email: string, password: string): Promise<User> {
    const user = await this.userService.findByEmail(email);

    if (user && compareSync(password, user.password)) {
      return user;
    }

    return null;
  }

  async login(login: LoginDto) {
    const user = await this.validateUser(login.email, login.password);

    if (!user) {
      throw new BadRequestException('Invalid credentials');
    }

    const accessToken = this.generateAccessToken(user);
    const refreshToken = await this.generateRefreshToken(user.id);

    return { accessToken, refreshToken };
  }

  async refreshAccessToken(user: User) {
    const accessToken = this.generateAccessToken(user);

    return { accessToken };
  }

  generateAccessToken(user: User) {
    const payload: JwtPayload = {
      sub: user.id,
      username: user.name,
      email: user.email,
    };
    const accessToken = this.jwtService.sign(payload, {
      expiresIn: process.env.JWT_EXPIRATION_AT,
      secret: process.env.JWT_SECRET,
    });

    return accessToken;
  }

  async generateRefreshToken(id: number) {
    const payload = {
      sub: id,
    };
    const refreshToken = await this.jwtService.sign(payload, {
      expiresIn: process.env.JWT_EXPIRATION_RT,
      secret: process.env.JWT_REFRESH_SECRET,
    });

    return refreshToken;
  }

  async verifyRefreshToken(refreshToken: string): Promise<JwtPayload> {
    try {
      const payload = await this.jwtService.verify(refreshToken, {
        secret: process.env.JWT_REFRESH_SECRET,
      });
      return payload;
    } catch (err) {
      throw new UnauthorizedException('Invalid refresh token');
    }
  }

  async verifyToken(token: string): Promise<boolean> {
    try {
      const decoded = await this.jwtService.verify(token, {
        secret: process.env.JWT_SECRET,
      });
      if (decoded) {
        return true && decoded;
      }
      return false;
    } catch (e) {
      return false;
    }
  }
}
