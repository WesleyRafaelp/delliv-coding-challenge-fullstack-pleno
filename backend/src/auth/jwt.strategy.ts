import { Injectable, UnauthorizedException } from '@nestjs/common';
import { PassportStrategy } from '@nestjs/passport';
import { ExtractJwt, Strategy } from 'passport-jwt';
import { UsersService } from 'src/users/users.service';
import { JwtPayload } from './interfaces/jwt-payload.interface';

@Injectable()
export class JwtStrategy extends PassportStrategy(Strategy) {
  constructor(private userService: UsersService) {
    super({
      jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
      secretOrKey: process.env.JWT_SECRET,
    });
  }

  async validate(payload: JwtPayload) {
    const user = await this.userService.validateUserId(payload.sub);

    if (!user) {
      throw new UnauthorizedException();
    }

    return user;
  }
}
