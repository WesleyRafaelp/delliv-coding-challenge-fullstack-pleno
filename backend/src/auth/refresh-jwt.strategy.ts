import { Injectable, UnauthorizedException } from '@nestjs/common';
import { PassportStrategy } from '@nestjs/passport';
import { ExtractJwt, Strategy } from 'passport-jwt';
import { UsersService } from 'src/users/users.service';
import { JwtPayload } from './interfaces/jwt-payload.interface';

@Injectable()
export class RefreshTokenStrategy extends PassportStrategy(
  Strategy,
  'jwt-refresh-token',
) {
  constructor(private userService: UsersService) {
    super({
      jwtFromRequest: ExtractJwt.fromExtractors([
        (request) =>
          request?.cookies?.REFRESH_TOKEN || request?.body?.refresh_token,
      ]),
      secretOrKey: process.env.JWT_REFRESH_SECRET,
    });
  }

  async validate(payload: JwtPayload): Promise<any> {
    const user = await this.userService.validateUserId(payload.sub);

    if (!user) {
      throw new UnauthorizedException();
    }

    return user;
  }
}
