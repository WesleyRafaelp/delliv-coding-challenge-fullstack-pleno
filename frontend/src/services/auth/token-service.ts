import { destroyCookie, parseCookies, setCookie } from "nookies";

export const tokenService = {
  save(accessToken: string, ctx = null) {
    setCookie(ctx, "ACCESS_TOKEN", accessToken, {
      maxAge: 30 * 24 * 60 * 60 * 1000,
      path: "/",
    });
  },
  get(ctx = null) {
    const cookies = parseCookies(ctx);
    return cookies["ACCESS_TOKEN"];
  },
  getCookies(ctx = null) {
    const cookies = parseCookies(ctx);
    return cookies;
  },
  delete(ctx = null) {
    destroyCookie(ctx, "ACCESS_TOKEN");

    destroyCookie(ctx, "REFRESH_TOKEN");
  },
};
