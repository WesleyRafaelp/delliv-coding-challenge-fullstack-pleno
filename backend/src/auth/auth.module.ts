import { Module } from '@nestjs/common';
import { JwtModule } from '@nestjs/jwt';
import { PassportModule } from '@nestjs/passport';
import { UsersModule } from 'src/users/users.module';
import { AuthController } from './auth.controller';
import { AuthService } from './auth.service';
import { JwtStrategy } from './jwt.strategy';
import { RefreshTokenStrategy } from './refresh-jwt.strategy';
import { RefreshTokenModule } from 'src/refresh-token/refresh-token.module';
import { PrismaService } from 'src/prisma.service';

@Module({
  imports: [
    UsersModule,
    PassportModule,
    RefreshTokenModule,
    PassportModule.register({ defaultStrategy: 'jwt' }),
    JwtModule.register({
      secret: process.env.JWT_SECRET,
      signOptions: { expiresIn: process.env.JWT_EXPIRATION_AT },
    }),
  ],
  controllers: [AuthController],
  providers: [AuthService, JwtStrategy, RefreshTokenStrategy, PrismaService],
  exports: [AuthService, JwtModule],
})
export class AuthModule {}
