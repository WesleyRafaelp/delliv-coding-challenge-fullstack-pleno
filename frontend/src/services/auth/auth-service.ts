import { HttpClient } from "../../infra/http-client/http-client";
import { tokenService } from "./token-service";

export const authService = {
  async login(user: any) {
    return await HttpClient(`${process.env.NEXT_PUBLIC_URL_API}/auth/login`, {
      method: "POST",
      body: user,
    }).then(async (res: any) => {
      if (!res.ok) throw new Error("Email or password invalid");
      
      const body = await res.body;

      tokenService.save(body.access_token);
    });
  },
  async getSession(ctx = null) {
    const token = tokenService.get(ctx);
    return await HttpClient(`${process.env.NEXT_PUBLIC_URL_API}/auth/session`, {
      method: "GET",
      headers: { "Authorization": `Bearer ${token}` },
      refresh: true,
      ctx,
    }).then(async (res: any) => {
      if (res.ok === false) throw new Error("Token invalid");
      const body = await res.body;
      const cookies = res.cookies;
    
      return { message: body.message, data: body.data };
    });
  },
  async logout(ctx = null) {
  
    const token = tokenService.get(ctx);
    return HttpClient(`${process.env.NEXT_PUBLIC_URL_API}/auth/logout`, {
      method: "POST",
      headers: { "Authorization": `Bearer ${token}` },
      refresh: true,
    }).then(async (res: any) => {
     
      if (!res.ok) throw new Error("Error logout");

      tokenService.delete(ctx);
      const body = await res.body;
  
      return { message: body.message };
    });
  },
};
