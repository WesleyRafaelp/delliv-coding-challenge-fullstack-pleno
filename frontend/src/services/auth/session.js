import { useEffect, useState } from "react";
import { authService } from "./auth-service";
import { useRouter } from "next/router";

export function useSession() {
  const [session, setSession] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);


  useEffect(() => {
    authService.getSession()
      .then((userSession) => {
        setSession(userSession)
      })
      .catch((err) => {
        setError(err)
      })
      .finally(() => {
        setLoading(false);
      });
  }, [])

  return {
    data: session,
    error,
    loading,
  }
}

export function withSessionHOC(WrappedComponent) {
  return function Wrapper(props) {
    const router = useRouter();
    const session = useSession();

    if (session.loading) {
      return ( <div className="fixed top-0 left-0 w-full h-full flex items-center justify-center bg-white opacity-75">
      <div className="loader ease-linear border rounded-full border-t-4 border-gray-900 h-12 w-12"></div>
    </div>);
    }

    if (!session.loading && session.error) {
      router.push('/?error=401');
    }

    const modifiedProps = {
      ...props,
      session: session.data?.data,
    };

    return <WrappedComponent { ...modifiedProps } />;
  };
}


export function withSession(functionCtx) {
  return async (ctx) => {
  
    try {
      const session = await authService.getSession(ctx);

      
      const modifiedContext = {
        ...ctx,
        req: {
          ...ctx.req,
          session,
        },
      };
      
      return functionCtx(modifiedContext);
    } catch (err) {
      
      console.error(err);
      return {
        redirect: {
          permanent: false,
          destination: "/?error=401",
        },
      };
    }
  };
}
