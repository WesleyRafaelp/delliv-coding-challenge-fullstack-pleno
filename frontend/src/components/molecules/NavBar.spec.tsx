import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { authService } from '@/services/auth/auth-service';
import { useRouter } from 'next/router';
import Navbar from './NavBar';

jest.mock('../../services/auth/auth-service', () => ({
  authService: {
    logout: jest.fn(),
  },
}));

jest.mock('next/router', () => ({
  ...jest.requireActual('next/router'),
  useRouter: jest.fn(() => ({ pathname: '/' })), 
}));

describe('Navbar Component', () => {

  it('calls authService.logout and redirects to home on button click', async () => {
    const mockPush = jest.fn().mockResolvedValue({});
    useRouter.mockReturnValue({
      push: mockPush,
    });
    render(<Navbar />);
    
    const logoutButton = screen.getByRole('button', { name: /sair/i });

    fireEvent.click(logoutButton);

    await waitFor(() => {
      expect(authService.logout).toHaveBeenCalled();
      expect(mockPush).toHaveBeenCalledWith('/');
    });
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<Navbar />);
    expect(asFragment()).toMatchSnapshot();
  });
});
