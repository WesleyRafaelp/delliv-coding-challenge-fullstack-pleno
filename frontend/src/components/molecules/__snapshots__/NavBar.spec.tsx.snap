// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Component matches snapshot 1`] = `
<DocumentFragment>
  <nav
    class="bg-gray-700 p-4 flex justify-between"
  >
    <ul
      class="flex items-center space-x-4 text-white"
    >
      <li>
        <a
          class="text-white-700
        hover:text-blue-500
        border-b-2 border-transparent
        transition-all duration-300 ease-in-out
        cursor-pointer
        font-semibold
        text-xl
      "
          href="/home"
        >
          Lista de Pedidos
        </a>
      </li>
      <li>
        <a
          class="text-white-700
        hover:text-blue-500
        border-b-2 border-transparent
        transition-all duration-300 ease-in-out
        cursor-pointer
        font-semibold
        text-xl
      "
          href="/create-order"
        >
          Criar Pedido
        </a>
      </li>
    </ul>
    <button
      class="bg-red-500 hover:bg-red-600 text-white font-bold py-2 px-4 rounded"
    >
      Sair
    </button>
  </nav>
</DocumentFragment>
`;
